app/services/editor-commands/editor-commands.ts

```diff
@@ -0,0 +1,273 @@
+            if (previousCommand.shouldCombine(instance)) {
+              previousCommand.combine(instance);
+              this.setCombineTimeout();
+              return;
+            }
+          }
+        }
+      }
+
+      this.setCombineTimeout();
+    }
+
+    this.undoHistory.push(instance);
+    this.PUSH_UNDO_METADATA({ description: instance.description });
+
+    return ret;
+  }
+
+  @shortcut('Ctrl+Z')
+  undo() {
+    if (this.state.operationInProgress) return;
+
+    const command = this.undoHistory.pop();
+    this.POP_UNDO_METADATA();
+
+    if (command) {
+      let ret: any;
+
+      try {
+        ret = command.rollback();
+      } catch (e) {
+        this.handleUndoRedoError(true, e);
+        return;
+      }
+
+      if (ret instanceof Promise) {
+        this.SET_OPERATION_IN_PROGRESS(true);
+        ret
+          .then(() => this.SET_OPERATION_IN_PROGRESS(false))
+          .catch(e => {
+            this.SET_OPERATION_IN_PROGRESS(false);
+            this.handleUndoRedoError(true, e);
+          });
+      }
+
+      this.redoHistory.push(command);
+      this.PUSH_REDO_METADATA({ description: command.description });
+    }
+  }
+
+  @shortcut('Ctrl+Y')
+  redo() {
+    if (this.state.operationInProgress) return;
+
+    const command = this.redoHistory.pop();
+    this.POP_REDO_METADATA();
+
+    if (command) {
+      let ret: any;
+
+      try {
+        ret = command.execute();
+      } catch (e) {
+        this.handleUndoRedoError(false, e);
+        return;
+      }
+
+      if (ret instanceof Promise) {
+        this.SET_OPERATION_IN_PROGRESS(true);
+        ret
+          .then(() => this.SET_OPERATION_IN_PROGRESS(false))
+          .catch(e => {
+            this.SET_OPERATION_IN_PROGRESS(false);
+            this.handleUndoRedoError(false, e);
+          });
+      }
+
+      this.undoHistory.push(command);
+      this.PUSH_UNDO_METADATA({ description: command.description });
+    }
+  }
+
+  private handleUndoRedoError(undo: boolean, e: any) {
+    console.error('Error performing undo operation', e);
+    electron.remote.dialog.showMessageBox({
+      title: 'Error',
+      message: `An error occurred while ${undo ? 'undoing' : 'redoing'} the operation.`,
+      type: 'error',
+    });
+    this.clear();
+  }
+
+  /**
+   * Clears the undo and redo histories entirely
+   */
+  private clear() {
+    this.undoHistory = [];
+    this.redoHistory = [];
+    this.CLEAR_UNDO_METADATA();
+    this.CLEAR_REDO_METADATA();
+  }
+
+  get nextUndo() {
+    return this.state.undoMetadata[this.state.undoMetadata.length - 1];
+  }
+
+  get nextUndoDescription() {
+    return this.nextUndo ? this.nextUndo.description : '';
+  }
+
+  get nextRedo() {
+    return this.state.redoMetadata[this.state.redoMetadata.length - 1];
+  }
+
+  get nextRedoDescription() {
+    return this.nextRedo ? this.nextRedo.description : '';
+  }
+
+  // Shortcuts for undo-able editor commands go here:
```

Comment:
why are these here now?

Reply:
I was originally trying not to have the selection service call into the editor commands service, but that broke down when I realized `remove` would be too difficult to port over.  I can probably move them back.